### PR TITLE
libc: newlib: make sbrk() thread-safe

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Concurrent use of this function could lead to corruption.
Add a semaphore to synchronize this.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Comment by andrewboie** [comment]
CI failure is broken in master branch, I'm opening a bug

**Review by dcpleung — APPROVED** [review]

File: lib/libc/newlib/libc-hooks.c
Comment (andyross) [inline lib/libc/newlib/libc-hooks.c]: Maybe a spinlock (which becomes an irq_lock on uniprocessor systems) would be a better fit here?  This is a fast operation with bounded runtime.  There's no real benefit to trying to block a process on it and significant extra code involved in sucking in the mutex implementation.

Comment (andrewboie) [reply lib/libc/newlib/libc-hooks.c]: This code has to run in user mode, so a spinlock is not appropriate.

Comment (andyross) [reply lib/libc/newlib/libc-hooks.c]: Futex then? :)

Just seems like the (comparatively very heavyweight, especially when done via syscalls) blocking mechanism is exactly what we don't want when synchronizing a tiny bit of deterministic code in the guts of the allocator.

Comment (andrewboie) [reply lib/libc/newlib/libc-hooks.c]: @andyross that's what sys_mutex is.
Implementation isn't completely done to use atomic ops, but that is what it is for. See #15138

Comment (andyross) [reply lib/libc/newlib/libc-hooks.c]: I thought sys_mutex was the priority inheriting thing (which we don't use much from elsewhere in the kernel, will suck in non-trivial extra code for all newlib targets, and which is known to have SMP bugs)?

You win.

**Review by andyross — APPROVED** [review]

File: lib/libc/newlib/libc-hooks.c
Comment (andrewboie) [reply lib/libc/newlib/libc-hooks.c]: > I thought sys_mutex was the priority inheriting thing 

It is. It is backed by a k_mutex on the kernel side. The design of sys mutex is to have the same semantics as FUTEX_LOCK_PI / FUTEX_UNLOCK_PI

> which is known to have SMP bugs

Which bugs are these?? We should label them high priority and fix them...

> will suck in non-trivial extra code for all newlib targets

What do you want here?
This code in its current state is broken.
Should I use a sys_sem instead?

**Comment by andrewboie** [comment]
Discussed some more offline, going to use a one-count sys_sem instead of a sys_mutex. Need to wait until the sys_sem patch is merged #16523

**Comment by andrewboie** [comment]
Changed to a sys_sem instead of a mutex.
Added a patch to add static sys_sem initializers.

**Comment by andrewboie** [comment]
> -:29: ERROR:MULTISTATEMENT_MACRO_USE_DO_WHILE: Macros with multiple statements should be enclosed in a do - while loop

This can't be done at toplevel.
K_SEM_DEFINE() does the same thing.
Spurious error!

File: include/sys/sem.h
Comment (wentongwu) [inline include/sys/sem.h]: I think we should also add BUILD_ASSERT here to check init count and count limit as sys_sem_init do.

**Review by wentongwu — DISMISSED** [review]

**Comment by kaidoho** [comment]
@andrewboie may I ask why you haven't used the ``malloc_lock`` / ``malloc_unlock`` hooks defined within newlib which guard the code around newlib's calls to sbrk? 

Perhaps a separate issue but closely related to this as malloc_lock / unlock require a reentrancy struct. The documentation within newlib's [reent.h](https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=newlib/libc/include/reent.h;h=2b01fbe8f329860fc7d76b681ea063375cb7eb72;hb=HEAD) instructs us to either provide reentrant versions of all syscalls and manage reentrancy structures ourself, or to keep a ``REENT_`` struct for each thread and provide newlib access to these structs via the ``__getreent()`` hook. As I can find neither an implementation of the ``*_r`` nor ``__getreent()`` - did I go terribly wrong somewhere or is the integration of newlib not completed yet?

**Comment by andrewboie** [comment]
> @andrewboie may I ask why you haven't used the malloc_lock / malloc_unlock hooks defined within newlib which guard the code around newlib's calls to sbrk?

I didn't know they existed.

> Perhaps a separate issue but closely related to this as malloc_lock / unlock require a reentrancy struct. The documentation within newlib's reent.h instructs us to either provide reentrant versions of all syscalls and manage reentrancy structures ourself, or to keep a REENT_ struct for each thread and provide newlib access to these structs via the __getreent() hook. As I can find neither an implementation of the *_r nor __getreent() - did I go terribly wrong somewhere or is the integration of newlib not completed yet?

@kaidoho Zephyr's newlib integration predates my involvement with the project -- I'm not the original author. I fixed the concurrency issue with sbrk() in a way I knew how, but if there's a better way to do some things please file a GH issue (or send patches)

**Comment by kaidoho** [comment]
Ok, will file an GH issue. Thx for your response @andrewboie

**Pull request merged**
